pep-0688.rst

```diff
@@ -0,0 +1,275 @@
+currently provides separate type aliases for writable and read-only
+buffers.
+
+However, in the C buffer protocol, these options cannot be
+queried directly on the type object. The only way to figure out
+whether an object supports a writable buffer is to actually
+ask for the buffer. For some types, such as ``memoryview``,
+whether the buffer is writable depends on the instance:
+some instances are read-only and others are not. As such, we propose to
+expose only whether a type implements the buffer protocol at
+all, not whether it supports more specific options such as
+writable buffers.
+
+Specification
+=============
+
+types.Buffer
+------------
+
+A new class, ``types.Buffer``, will be added. It cannot be instantiated or
+subclassed, but supports the ``__instancecheck__`` and
+``__subclasscheck__`` hooks.  In CPython, these will check for the presence of the
+``bf_getbuffer`` slot in the type object:
+
+.. code-block:: pycon
+
+   >>> from types import Buffer
+   >>> isinstance(b"xy", Buffer)
+   True
+   >>> issubclass(bytes, Buffer)
+   True
+   >>> issubclass(memoryview, Buffer)
+   True
+   >>> isinstance("xy", Buffer)
+   False
+   >>> issubclass(str, Buffer)
+   False
+
+The new class can also be used in type annotations:
+
+.. code-block:: python
+
+   def need_buffer(b: Buffer) -> memoryview:
+       return memoryview(b)
+
+   need_buffer(b"xy")  # ok
+   need_buffer("xy")  # rejected by static type checkers
+
+Usage in stub files
+-------------------
+
+For static typing purposes, types defined in C extensions usually
+require stub files, as :pep:`described in PEP 484 <484#stub-files>`.
+In stub files, ``types.Buffer`` may be used as a base class to
+indicate that a class implements the buffer protocol.
+
+For example, ``bytes`` may be declared as follows in a stub:
+
+.. code-block:: python
+
+   class bytes(types.Buffer, Sequence[int]):
+       def decode(self, ...): ...
+       ...
+
+The ``types.Buffer`` class does not require any special treatment
+in type checkers.
+
+Equivalent for older Python versions
+------------------------------------
+
+New typing features are usually backported to older Python versions
+in the `typing_extensions <https://pypi.org/project/typing-extensions/>`_
+package. Because the buffer protocol
+is accessible only in C, ``types.Buffer`` cannot be implemented
+in a pure-Python package like ``typing_extensions``. As a temporary
+workaround, a ``typing_extensions.Buffer``
+`Abstract Base Class <Buffer ABC_>`__ will be provided on Python versions
+that do not have ``types.Buffer`` available. For the benefit of
+static type checkers, ``typing_extensions.Buffer`` can be used as
+a base class in stubs to mark types as supporting the buffer protocol.
+For runtime uses, the ``ABC.register`` API can be used to register
+buffer classes with ``typing_extensions.Buffer``. When
+``types.Buffer`` is available, ``typing_extensions`` should simply
+re-export it.
```

Comment:
Do we need to state this in the PEP? -- isn't this something of a `typing_extensions` implementation detail?

Reply:
This is important if people do `typing_extensions.Buffer.register(MyFancyBuffer)`. I'll put a note to make that explicit.